### PR TITLE
add method for data migration

### PR DESCRIPTION
**Pull request by YuliiaKovalova**
## Problem
Add method for handling first-run data migration
#6334

## Solution
Add method for initial data migration on first package request.

### CODE REVIEW

File: src/Microsoft.TemplateEngine.Abstractions/IEngineEnvironmentSettings.cs [hunk 1]
```diff
@@ -35,5 +35,10 @@ public interface IEngineEnvironmentSettings : IDisposable
         /// Component manager for this instance of settings.
         /// </summary>
         IComponentManager Components { get; }
+
+        /// <summary>
+        /// Flag specifies if tab completion mode is enabled.
+        /// </summary>
+        bool TabCompletionMode { get; }
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Abstractions/IEngineEnvironmentSettings.cs:42]: This looks as too specific for `IEngineEnvironmentSettings`, also causes breaking changes.
Please consider moving to `TemplatePackageManager` instead or maybe to `IEnvironment`.

Comment (YuliiaKovalova) [reply src/Microsoft.TemplateEngine.Abstractions/IEngineEnvironmentSettings.cs]: done

File: src/Microsoft.TemplateEngine.Edge/DefaultEnvironment.cs [hunk 2]
```diff
@@ -18,11 +18,12 @@ public class DefaultEnvironment : IEnvironment
         private const int DefaultBufferWidth = 160;
         private readonly IReadOnlyDictionary<string, string> _environmentVariables;
 
-        public DefaultEnvironment() : this(FetchEnvironmentVariables())
+        public DefaultEnvironment() : this(tabCompletionMode: false, FetchEnvironmentVariables())
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Edge/DefaultEnvironment.cs:21]: please add it as the last parameter and optional to ensure backwards compatibility.

File: src/Microsoft.TemplateEngine.Edge/BuiltInManagedProvider/GlobalSettings.cs [hunk 3]
```diff
@@ -157,6 +163,75 @@ public async Task SetInstalledTemplatePackagesAsync(IReadOnlyList<TemplatePackag
...
+                return;
+            }
+
+            _logger.LogWarning(LocalizableStrings.GlobalSettingsTemplatePackageProvider_DataMigrationWarning);
+
+            string sourceFeedKey = "NuGetSource";
+            string packageIdKey = "PackageId";
+            string versionKey = "Version";
+            string ownersKey = "Owners";
+            string trustedKey = "Trusted";
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Edge/BuiltInManagedProvider/GlobalSettings.cs:179]: It's better to use those https://github.com/dotnet/templating/blob/6536e8d9b4ae1e70047cf4c254c352db75b99903/src/Microsoft.TemplateEngine.Edge/Installers/NuGet/NuGetManagedTemplatePackage.cs#L16-L22, just make them internal.

File: src/Microsoft.TemplateEngine.Edge/BuiltInManagedProvider/GlobalSettings.cs [hunk 4]
```diff
@@ -157,6 +163,75 @@ public async Task SetInstalledTemplatePackagesAsync(IReadOnlyList<TemplatePackag
...
         }
 
+        public async Task MigrateCacheMetadataAsync(CancellationToken cancellationToken)
+        {
+            if (!CanApplyFirstRunDataMigration())
+            {
+                return;
+            }
+
+            _logger.LogWarning(LocalizableStrings.GlobalSettingsTemplatePackageProvider_DataMigrationWarning);
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Edge/BuiltInManagedProvider/GlobalSettings.cs:173]:
```suggestion
            _logger.LogInformation(LocalizableStrings.GlobalSettingsTemplatePackageProvider_DataMigrationWarning);
```

File: src/Microsoft.TemplateEngine.Edge/BuiltInManagedProvider/GlobalSettings.cs [hunk 5]
```diff
@@ -157,6 +163,75 @@ public async Task SetInstalledTemplatePackagesAsync(IReadOnlyList<TemplatePackag
...
+                    else
+                    {
+                        updatedPackages.Add(package);
+                    }
+                }
+
+                using var stream = _environmentSettings.Host.FileSystem.CreateFile(_paths.FirstRunCookie);
+
+                // cleanup existing cache file
+                await SetInstalledTemplatePackagesAsync(new List<TemplatePackageData>(), cancellationToken).ConfigureAwait(false);
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Edge/BuiltInManagedProvider/GlobalSettings.cs:220]: i don't think this is needed, just update the packages list directly. This method actually writes the file and it's performance impact.

File: src/Microsoft.TemplateEngine.Edge/BuiltInManagedProvider/GlobalSettings.cs [hunk 6]
```diff
@@ -157,6 +163,75 @@ public async Task SetInstalledTemplatePackagesAsync(IReadOnlyList<TemplatePackag
             throw new InvalidOperationException();
         }
 
+        public async Task MigrateCacheMetadataAsync(CancellationToken cancellationToken)
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Edge/BuiltInManagedProvider/GlobalSettings.cs:166]: whole method should be locked as here https://github.com/dotnet/templating/blob/6536e8d9b4ae1e70047cf4c254c352db75b99903/src/Microsoft.TemplateEngine.Edge/BuiltInManagedProvider/GlobalSettingsTemplatePackageProvider.cs#L171
before cookie is checked, so migration is not performed twice.

File: src/Microsoft.TemplateEngine.Edge/BuiltInManagedProvider/GlobalSettings.cs [hunk 7]
```diff
@@ -157,6 +163,75 @@ public async Task SetInstalledTemplatePackagesAsync(IReadOnlyList<TemplatePackag
...
+                        extendedPackageDetails[ownersKey] = updatedPackageMetadata.Owners;
+                        updatedPackages.Add(new TemplatePackageData(package.InstallerId, package.MountPointUri, package.LastChangeTime, extendedPackageDetails));
+                    }
+                    else
+                    {
+                        updatedPackages.Add(package);
+                    }
+                }
+
+                using var stream = _environmentSettings.Host.FileSystem.CreateFile(_paths.FirstRunCookie);
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Edge/BuiltInManagedProvider/GlobalSettings.cs:217]: this should be the last operation. Moreover, if migration critically failed, it should be also written not to re-attempt it.

File: src/Microsoft.TemplateEngine.Edge/BuiltInManagedProvider/GlobalSettings.cs [hunk 8]
```diff
@@ -157,6 +163,75 @@ public async Task SetInstalledTemplatePackagesAsync(IReadOnlyList<TemplatePackag
...
+                // cleanup existing cache file
+                await SetInstalledTemplatePackagesAsync(new List<TemplatePackageData>(), cancellationToken).ConfigureAwait(false);
+                await SetInstalledTemplatePackagesAsync(updatedPackages, cancellationToken).ConfigureAwait(false);
+            }
+            catch
+            {
+                if (_environmentSettings.Host.FileSystem.FileExists(_paths.FirstRunCookie))
+                {
+                    _environmentSettings.Host.FileSystem.FileDelete(_paths.FirstRunCookie);
+                }
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Edge/BuiltInManagedProvider/GlobalSettings.cs:228]: if migration critically failed, the file should be written not to re-attempt it.

File: src/Microsoft.TemplateEngine.Edge/BuiltInManagedProvider/GlobalSettings.cs [hunk 9]
```diff
@@ -157,6 +163,75 @@ public async Task SetInstalledTemplatePackagesAsync(IReadOnlyList<TemplatePackag
             throw new InvalidOperationException();
         }
 
+        public async Task MigrateCacheMetadataAsync(CancellationToken cancellationToken)
+        {
+            if (!CanApplyFirstRunDataMigration())
+            {
+                return;
+            }
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Edge/BuiltInManagedProvider/GlobalSettings.cs:171]: Where the cookie file is checked?

File: src/Microsoft.TemplateEngine.Edge/BuiltInManagedProvider/GlobalSettingsTemplatePackageProvider.cs [hunk 10]
```diff
@@ -127,6 +127,8 @@ public async Task<IReadOnlyList<InstallResult>> InstallAsync(IEnumerable<Install
             }
 
             using var disposable = await _globalSettings.LockAsync(cancellationToken).ConfigureAwait(false);
+            await _globalSettings.MigrateCacheMetadataAsync(cancellationToken).ConfigureAwait(false);
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Edge/BuiltInManagedProvider/GlobalSettingsTemplatePackageProvider.cs:130]: better to run it in global settings itself in single place.

File: src/Microsoft.TemplateEngine.Edge/BuiltInManagedProvider/GlobalSettings.cs (hunk 7 above)
Comment (vlada-shubina) [reply src/Microsoft.TemplateEngine.Edge/BuiltInManagedProvider/GlobalSettings.cs:217]: Also consider separate name for the global settings migration cookie.

File: src/Microsoft.TemplateEngine.Edge/BuiltInManagedProvider/GlobalSettings.cs [hunk 11]
```diff
@@ -157,6 +163,75 @@ public async Task SetInstalledTemplatePackagesAsync(IReadOnlyList<TemplatePackag
...
+
+            string sourceFeedKey = "NuGetSource";
+            string packageIdKey = "PackageId";
+            string versionKey = "Version";
+            string ownersKey = "Owners";
+            string trustedKey = "Trusted";
+
+            var updatedPackages = new List<TemplatePackageData>();
+            var metadataDownloader = new NuGetApiPackageManager(_environmentSettings);
+            var packages = await GetInstalledTemplatePackagesAsync(cancellationToken).ConfigureAwait(false);
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Edge/BuiltInManagedProvider/GlobalSettings.cs:183]: the idea of those method is get you the list that is modified directly if needed and then written using `SetInstalledTemplatePackagesAsync`

File: src/Microsoft.TemplateEngine.Edge/BuiltInManagedProvider/GlobalSettings.cs [hunk 12]
```diff
@@ -157,6 +163,75 @@ public async Task SetInstalledTemplatePackagesAsync(IReadOnlyList<TemplatePackag
...
+                        var updatedPackageMetadata = await metadataDownloader.GetPackageMetadataAsync(
+                            identifier,
+                            version,
+                            packageSource,
+                            CancellationToken.None).ConfigureAwait(false);
+
+                        var extendedPackageDetails = package.Details.ToDictionary(kvp => kvp.Key, kvp => kvp.Value);
+                        extendedPackageDetails[trustedKey] = updatedPackageMetadata.Trusted.ToString() ?? false.ToString();
+                        extendedPackageDetails[ownersKey] = updatedPackageMetadata.Owners;
+                        updatedPackages.Add(new TemplatePackageData(package.InstallerId, package.MountPointUri, package.LastChangeTime, extendedPackageDetails));
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Edge/BuiltInManagedProvider/GlobalSettings.cs:209]: update the `package` directly

File: src/Microsoft.TemplateEngine.Edge/BuiltInManagedProvider/GlobalSettings.cs [hunk 13]
```diff
@@ -157,6 +163,75 @@ public async Task SetInstalledTemplatePackagesAsync(IReadOnlyList<TemplatePackag
...
+            _logger.LogWarning(LocalizableStrings.GlobalSettingsTemplatePackageProvider_DataMigrationWarning);
+
+            string sourceFeedKey = "NuGetSource";
+            string packageIdKey = "PackageId";
+            string versionKey = "Version";
+            string ownersKey = "Owners";
+            string trustedKey = "Trusted";
+
+            var updatedPackages = new List<TemplatePackageData>();
+            var metadataDownloader = new NuGetApiPackageManager(_environmentSettings);
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Edge/BuiltInManagedProvider/GlobalSettings.cs:182]: consider using already created package manager and passing it as parameter.

**Review by vlada-shubina — COMMENTED** [review]
Just checked the migration part so far.

File: tools/Microsoft.TemplateEngine.Authoring.CLI/Commands/validate/ValidateCommand.cs [hunk 14]
```diff
@@ -54,8 +54,8 @@ protected override async Task<int> ExecuteAsync(ValidateCommandArgs args, ILogge
         private IEngineEnvironmentSettings SetupSettings(ILoggerFactory loggerFactory)
         {
             var builtIns = new List<(Type InterfaceType, IIdentifiedComponent Instance)>();
-            builtIns.AddRange(Microsoft.TemplateEngine.Orchestrator.RunnableProjects.Components.AllComponents);
-            builtIns.AddRange(Microsoft.TemplateEngine.Edge.Components.AllComponents);
+            builtIns.AddRange(Orchestrator.RunnableProjects.Components.AllComponents);
+            builtIns.AddRange(Components.AllComponents);
```
Comment (vlada-shubina) [inline tools/Microsoft.TemplateEngine.Authoring.CLI/Commands/validate/ValidateCommand.cs:58]: This change is not needed for the purpose of the PR.

**Review by vlada-shubina — CHANGES_REQUESTED** [review]
The migration part should be locked. 
The package should be only attempted to be updated from the source if was installed. 
What happens if the update of the package fails (i.e. `GetPackageMetadataAsync` throws an exception)?

**Pull request closed without merging**